controllers/tls_workflow.go

```diff
@@ -0,0 +1,147 @@
+package controllers
+
+import (
+	"github.com/cert-manager/cert-manager/pkg/apis/certmanager"
+	certmanagerv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	"github.com/kuadrant/policy-machinery/controller"
+	"github.com/kuadrant/policy-machinery/machinery"
+	"github.com/samber/lo"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	gwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+
+	kuadrantv1alpha1 "github.com/kuadrant/kuadrant-operator/api/v1alpha1"
+)
+
+var (
+	CertManagerCertificatesResource  = certmanagerv1.SchemeGroupVersion.WithResource("certificates")
+	CertManagerIssuersResource       = certmanagerv1.SchemeGroupVersion.WithResource("issuers")
+	CertMangerClusterIssuersResource = certmanagerv1.SchemeGroupVersion.WithResource("clusterissuers")
+
+	CertManagerCertificateKind   = schema.GroupKind{Group: certmanager.GroupName, Kind: certmanagerv1.CertificateKind}
+	CertManagerIssuerKind        = schema.GroupKind{Group: certmanager.GroupName, Kind: certmanagerv1.IssuerKind}
+	CertManagerClusterIssuerKind = schema.GroupKind{Group: certmanager.GroupName, Kind: certmanagerv1.ClusterIssuerKind}
+)
+
+func NewTLSPolicyWorkflow(client *dynamic.DynamicClient) *controller.Workflow {
+	return &controller.Workflow{
+		Precondition:  NewValidateTLSPoliciesValidatorReconciler().Validate,
+		Postcondition: NewTLSPolicyStatusUpdaterReconciler(client).UpdateStatus,
+	}
+}
+
+func LinkGatewayToCertificateFunc(objs controller.Store) machinery.LinkFunc {
+	gateways := lo.Map(objs.FilterByGroupKind(machinery.GatewayGroupKind), controller.ObjectAs[*gwapiv1.Gateway])
+
+	return machinery.LinkFunc{
+		From: machinery.GatewayGroupKind,
+		To:   CertManagerCertificateKind,
+		Func: func(child machinery.Object) []machinery.Object {
+			o := child.(*controller.RuntimeObject)
+			cert := o.Object.(*certmanagerv1.Certificate)
+
+			gateway, ok := lo.Find(gateways, func(g *gwapiv1.Gateway) bool {
+				for _, l := range g.Spec.Listeners {
```

Comment:
I wonder if these objects should be liked to the gateway listeners instead of the gateways as a whole. If not now, in the near future.

With `sectionName` introduced to the TLSPolicy, the policies themselves may be linked to the entire gateway or to a specific listener only. I suppose that could make a difference in the decision of the number of CSRs that are created for each policy resource – one with multiple CNs or one per listener.

Either way, the case of a certificate linked to a individual listener obviously exists. It's actually the default today as we currently do not create CSRs with multiple CNs IIRC. As I can tell from this loop, the listener is definitely the driver to decide if the object is linked or not anyway.

WDYT?

Interested in your thoughts too @maleck13.

Reply:
Yes I think having TLSPolicy work against individual listeners is the way to go. I think we can think of it similar to DNS 
- Each listener is evaluated and the DNSPolicy reconciled for that listener
- Each listener has its status updated in the gateway with enforced  / accpeted conditions
- Probably remove the overall gateway status. 

^ some of this may be out of scope for now

Reply:
Sounds good, I'll see can I do the linking to certificates instead as part of this PR. If not, I'll do that change with the refactor of main reconcile logic 👍